File: ros2_ws/src/perception/lantern_detector_pkg/lantern_detector_pkg/lantern_detector_node.py
import math
from dataclasses import dataclass
from pathlib import Path
from typing import List, Optional, Tuple

import cv2
import numpy as np
import rclpy
from cv_bridge import CvBridge
from geometry_msgs.msg import PoseStamped
from rclpy.node import Node
from sensor_msgs.msg import CameraInfo, Image
from tf2_ros import Buffer, TransformListener
import tf2_geometry_msgs  # noqa: F401
from message_filters import ApproximateTimeSynchronizer, Subscriber


@dataclass
class Track:
    position: np.ndarray
    observations: int
    published: bool = False


class LanternDetectorNode(Node):
    def __init__(self) -> None:
        super().__init__("lantern_detector")

        self.declare_parameter("semantic_topic", "/realsense/semantic/image_rect_raw")
        self.declare_parameter("depth_topic", "/realsense/depth/image")
        self.declare_parameter("camera_info_topic", "/realsense/semantic/camera_info")
        self.declare_parameter("world_frame", "world")
        self.declare_parameter("body_frame", "body")
        self.declare_parameter("camera_offset", [0.1, 0.0, 0.0])
        self.declare_parameter("min_area", 10.0)
        self.declare_parameter("depth_window", 5)
        self.declare_parameter("hsv_lower", [20, 90, 90])
        self.declare_parameter("hsv_upper", [70, 255, 255])
        self.declare_parameter("gating_distance", 2.0)
        self.declare_parameter("min_observations", 5)
        self.declare_parameter("tf_timeout_s", 0.2)
        self.declare_parameter("use_latest_tf_on_extrapolation", True)

        self.bridge = CvBridge()
        self.camera_info: Optional[CameraInfo] = None
        self.tracks: List[Track] = []

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        self.detections_pub = self.create_publisher(PoseStamped, "/detected_lanterns", 10)

        semantic_sub = Subscriber(self, Image, self.get_parameter("semantic_topic").value)
        depth_sub = Subscriber(self, Image, self.get_parameter("depth_topic").value)
        info_sub = Subscriber(self, CameraInfo, self.get_parameter("camera_info_topic").value)

        sync = ApproximateTimeSynchronizer([semantic_sub, depth_sub, info_sub], 10, 0.1)
        sync.registerCallback(self.on_images)

        self.get_logger().info("Lantern detector initialized")

    def on_images(self, semantic_msg: Image, depth_msg: Image, info_msg: CameraInfo) -> None:
        self.camera_info = info_msg
        try:
            semantic = self.bridge.imgmsg_to_cv2(semantic_msg, desired_encoding="bgr8")
            depth = self.bridge.imgmsg_to_cv2(depth_msg)
        except Exception as exc:
            self.get_logger().warn(f"Failed to convert images: {exc}")
            return

        mask = self.create_yellow_mask(semantic)
        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

        for contour in contours:
            area = cv2.contourArea(contour)
            if area < float(self.get_parameter("min_area").value):
                continue

            centroid = self.compute_centroid(contour)
            if centroid is None:
                continue

            depth_m = self.sample_depth(depth, centroid)
            if depth_m is None or math.isnan(depth_m) or depth_m <= 0.0:
                continue

            camera_point = self.project_to_camera(centroid, depth_m)
            if camera_point is None:
                continue

            world_point = self.camera_to_world(
                camera_point,
                semantic_msg.header.stamp,
            )
            if world_point is None:
                continue

            self.update_tracks(world_point)

        self.publish_tracks(semantic_msg.header.stamp)

    def create_yellow_mask(self, image: np.ndarray) -> np.ndarray:
        hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
        lower = np.array(self.get_parameter("hsv_lower").value, dtype=np.uint8)
        upper = np.array(self.get_parameter("hsv_upper").value, dtype=np.uint8)
        mask = cv2.inRange(hsv, lower, upper)
        return mask

    @staticmethod
    def compute_centroid(contour: np.ndarray) -> Optional[Tuple[int, int]]:
        moments = cv2.moments(contour)
        if moments["m00"] == 0:
            return None
        u = int(moments["m10"] / moments["m00"])
        v = int(moments["m01"] / moments["m00"])
        return u, v

    def sample_depth(self, depth: np.ndarray, centroid: Tuple[int, int]) -> Optional[float]:
        u, v = centroid
        window = int(self.get_parameter("depth_window").value)
        half = max(1, window // 2)
        h, w = depth.shape[:2]
        u_min, u_max = max(0, u - half), min(w, u + half + 1)
        v_min, v_max = max(0, v - half), min(h, v + half + 1)
        patch = depth[v_min:v_max, u_min:u_max].astype(np.float32)
        if patch.size == 0:
            return None
        if depth.dtype == np.uint16:
            patch = patch / 1000.0
        median = float(np.median(patch))
        return median

    def project_to_camera(self, centroid: Tuple[int, int], depth_m: float) -> Optional[np.ndarray]:
        if self.camera_info is None:
            return None
        u, v = centroid
        k = self.camera_info.k
        fx, fy = k[0], k[4]
        cx, cy = k[2], k[5]
        if fx == 0 or fy == 0:
            return None
        x = (u - cx) / fx * depth_m
        y = (v - cy) / fy * depth_m
        z = depth_m
        return np.array([x, y, z], dtype=np.float64)

    def camera_to_world(self, camera_point: np.ndarray, stamp) -> Optional[np.ndarray]:
        pose_frame = self.get_parameter("body_frame").value
        offset = np.array(self.get_parameter("camera_offset").value, dtype=np.float64)
        body_point = self.camera_to_body(camera_point) + offset
        pose = PoseStamped()
        pose.header.stamp = stamp
        timeout_s = float(self.get_parameter("tf_timeout_s").value)
        use_latest = bool(self.get_parameter("use_latest_tf_on_extrapolation").value)
        pose.header.frame_id = pose_frame
        pose.pose.position.x = float(body_point[0])
        pose.pose.position.y = float(body_point[1])
        pose.pose.position.z = float(body_point[2])
        pose.pose.orientation.w = 1.0
        world_pose = self.transform_pose(pose, timeout_s, use_latest)
        if world_pose is None:
            return None

        return np.array(
            [
                world_pose.pose.position.x,
                world_pose.pose.position.y,
                world_pose.pose.position.z,
            ],
            dtype=np.float64,
        )

    @staticmethod
    def camera_to_body(camera_point: np.ndarray) -> np.ndarray:
        rotation = np.array(
            [
                [0.0, 0.0, 1.0],
                [1.0, 0.0, 0.0],
                [0.0, 1.0, 0.0],
            ],
            dtype=np.float64,
        )
        return rotation @ camera_point

    def transform_pose(self, pose: PoseStamped, timeout_s: float, use_latest: bool) -> Optional[PoseStamped]:
        try:
             return self.tf_buffer.transform(
                pose,
                self.get_parameter("world_frame").value,
                timeout=rclpy.duration.Duration(seconds=timeout_s),
            )
        except Exception as exc:
            if use_latest and "extrapolation into the future" in str(exc).lower():
                try:
                    pose.header.stamp = rclpy.time.Time()
                    return self.tf_buffer.transform(
                        pose,
                        self.get_parameter("world_frame").value,
                        timeout=rclpy.duration.Duration(seconds=timeout_s),
                    )
                except Exception as retry_exc:
                    self.get_logger().warn(f"TF transform failed: {retry_exc}")
                    return None
            self.get_logger().warn(f"TF transform failed: {exc}")
            return None
        
    def update_tracks(self, world_point: np.ndarray) -> None:
        gating_distance = float(self.get_parameter("gating_distance").value)
        best_track = None
        best_distance = None
        for track in self.tracks:
            distance = float(np.linalg.norm(track.position - world_point))
            if distance <= gating_distance and (best_distance is None or distance < best_distance):
                best_track = track
                best_distance = distance

        if best_track is None:
            self.tracks.append(Track(position=world_point, observations=1))
            return

        count = best_track.observations
        best_track.position = (best_track.position * count + world_point) / (count + 1)
        best_track.observations += 1

    def publish_tracks(self, stamp) -> None:
        min_observations = int(self.get_parameter("min_observations").value)
        world_frame = self.get_parameter("world_frame").value
        for track in self.tracks:
            if track.observations < min_observations:
                continue
            if track.published:
                continue
            pose = PoseStamped()
            pose.header.stamp = stamp
            pose.header.frame_id = world_frame
            pose.pose.position.x = float(track.position[0])
            pose.pose.position.y = float(track.position[1])
            pose.pose.position.z = float(track.position[2])
            pose.pose.orientation.w = 1.0
            self.detections_pub.publish(pose)
            track.published = True

def main() -> None:
    rclpy.init()
    node = LanternDetectorNode()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


class LanternDetectionLoggerNode(Node):
    def __init__(self) -> None:
        super().__init__("lantern_detection_logger")

        self.declare_parameter("detections_topic", "/detected_lanterns")
        self.declare_parameter("output_file", "lantern_detections.txt")

        self.detections: List[PoseStamped] = []
        self.create_subscription(
            PoseStamped,
            self.get_parameter("detections_topic").value,
            self.on_detection,
            10,
        )

        self.get_logger().info("Lantern detection logger initialized")

    def on_detection(self, msg: PoseStamped) -> None:
        self.detections.append(msg)

    def destroy_node(self) -> bool:
        self.write_detections_to_file()
        return super().destroy_node()

    def write_detections_to_file(self) -> None:
        output_file = Path(self.get_parameter("output_file").value)
        lines = ["frame_id,x,y,z"]
        for detection in self.detections:
            position = detection.pose.position
            lines.append(
                f"{detection.header.frame_id},{position.x:.3f},{position.y:.3f},{position.z:.3f}"
            )
        try:
            output_file.parent.mkdir(parents=True, exist_ok=True)
            output_file.write_text("\n".join(lines) + "\n")
            self.get_logger().info(f"Wrote lantern detections to {output_file}")
        except Exception as exc:
            self.get_logger().warn(f"Failed to write lantern detections: {exc}")


def logger_main() -> None:
    rclpy.init()
    node = LanternDetectionLoggerNode()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2_ws/src/perception/lantern_detector_pkg/src/lantern_detector_node.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <cmath>
#include <functional>
#include <memory>
#include <optional>
#include <string>
#include <utility>
#include <vector>

#include "cv_bridge/cv_bridge.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "message_filters/subscriber.h"
#include "message_filters/sync_policies/approximate_time.h"
#include "message_filters/synchronizer.h"
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/camera_info.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"
#include "tf2/time.h"

#include <opencv2/imgproc.hpp>

namespace {
struct Track {
  std::array<double, 3> position{};
  int observations = 0;
  bool published = false;
};

std::optional<std::pair<int, int>> ComputeCentroid(const std::vector<cv::Point> & contour) {
  cv::Moments moments = cv::moments(contour);
  if (moments.m00 == 0.0) {
    return std::nullopt;
  }
  int u = static_cast<int>(moments.m10 / moments.m00);
  int v = static_cast<int>(moments.m01 / moments.m00);
  return std::make_pair(u, v);
}

std::optional<double> MedianOf(std::vector<float> values) {
  if (values.empty()) {
    return std::nullopt;
  }
  const size_t mid = values.size() / 2;
  std::nth_element(values.begin(), values.begin() + mid, values.end());
  if (values.size() % 2 == 1) {
    return static_cast<double>(values[mid]);
  }
  const float upper = values[mid];
  std::nth_element(values.begin(), values.begin() + mid - 1, values.end());
  const float lower = values[mid - 1];
  return 0.5 * (static_cast<double>(lower) + static_cast<double>(upper));
}

std::string ToLower(std::string value) {
  std::transform(value.begin(), value.end(), value.begin(), [](unsigned char c) {
    return static_cast<char>(std::tolower(c));
  });
  return value;
}
}  // namespace

class LanternDetectorNode : public rclcpp::Node {
public:
  LanternDetectorNode()
  : Node("lantern_detector") {
    declare_parameter<std::string>("semantic_topic", "/realsense/semantic/image_rect_raw");
    declare_parameter<std::string>("depth_topic", "/realsense/depth/image");
    declare_parameter<std::string>("camera_info_topic", "/realsense/semantic/camera_info");
    declare_parameter<std::string>("world_frame", "world");
    declare_parameter<std::string>("body_frame", "body");
    declare_parameter<std::vector<double>>("camera_offset", {0.1, 0.0, 0.0});
    declare_parameter<double>("min_area", 5.0);
    declare_parameter<int>("depth_window", 5);
    declare_parameter<std::vector<int64_t>>("hsv_lower", {20, 70, 70});
    declare_parameter<std::vector<int64_t>>("hsv_upper", {70, 255, 255});
    declare_parameter<double>("gating_distance", 2.0);
    declare_parameter<int>("min_observations", 5);
    declare_parameter<double>("tf_timeout_s", 0.2);
    declare_parameter<bool>("use_latest_tf_on_extrapolation", true);

    detections_pub_ = create_publisher<geometry_msgs::msg::PoseStamped>("/detected_lanterns", 10);

    const auto semantic_topic = get_parameter("semantic_topic").as_string();
    const auto depth_topic = get_parameter("depth_topic").as_string();
    const auto camera_info_topic = get_parameter("camera_info_topic").as_string();

    semantic_sub_.subscribe(this, semantic_topic);
    depth_sub_.subscribe(this, depth_topic);
    info_sub_.subscribe(this, camera_info_topic);

    sync_ = std::make_shared<Sync>(SyncPolicy(10), semantic_sub_, depth_sub_, info_sub_);
    sync_->setMaxIntervalDuration(rclcpp::Duration::from_seconds(0.05));
    sync_->registerCallback(std::bind(
      &LanternDetectorNode::OnImages, this, std::placeholders::_1, std::placeholders::_2,
      std::placeholders::_3));

    tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

    RCLCPP_INFO(get_logger(), "Lantern detector initialized");
  }

private:
  using SyncPolicy = message_filters::sync_policies::ApproximateTime<
    sensor_msgs::msg::Image, sensor_msgs::msg::Image, sensor_msgs::msg::CameraInfo>;
  using Sync = message_filters::Synchronizer<SyncPolicy>;

  void OnImages(
    const sensor_msgs::msg::Image::ConstSharedPtr & semantic_msg,
    const sensor_msgs::msg::Image::ConstSharedPtr & depth_msg,
    const sensor_msgs::msg::CameraInfo::ConstSharedPtr & info_msg) {
    camera_info_ = info_msg;

    cv::Mat semantic;
    cv::Mat depth;
    try {
      semantic = cv_bridge::toCvShare(semantic_msg, "bgr8")->image;
      depth = cv_bridge::toCvShare(depth_msg)->image;
    } catch (const cv_bridge::Exception & exc) {
      RCLCPP_WARN(get_logger(), "Failed to convert images: %s", exc.what());
      return;
    }

    cv::Mat mask = CreateYellowMask(semantic);
    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

    for (const auto & contour : contours) {
      const double area = cv::contourArea(contour);
      if (area < get_parameter("min_area").as_double()) {
        continue;
      }

      auto centroid = ComputeCentroid(contour);
      if (!centroid) {
        continue;
      }

      auto depth_m = SampleDepth(depth, *centroid);
      if (!depth_m || std::isnan(*depth_m) || *depth_m <= 0.0) {
        continue;
      }

      auto camera_point = ProjectToCamera(*centroid, *depth_m);
      if (!camera_point) {
        continue;
      }

      auto world_point = CameraToWorld(*camera_point, semantic_msg->header.stamp);
      if (!world_point) {
        continue;
      }

      UpdateTracks(*world_point);
    }

    PublishTracks(semantic_msg->header.stamp);
  }

  cv::Mat CreateYellowMask(const cv::Mat & image) {
    cv::Mat hsv;
    cv::cvtColor(image, hsv, cv::COLOR_BGR2HSV);
    const auto lower_param = get_parameter("hsv_lower").as_integer_array();
    const auto upper_param = get_parameter("hsv_upper").as_integer_array();
    cv::Scalar lower(
      static_cast<int>(lower_param.at(0)),
      static_cast<int>(lower_param.at(1)),
      static_cast<int>(lower_param.at(2)));
    cv::Scalar upper(
      static_cast<int>(upper_param.at(0)),
      static_cast<int>(upper_param.at(1)),
      static_cast<int>(upper_param.at(2)));
    cv::Mat mask;
    cv::inRange(hsv, lower, upper, mask);
    return mask;
  }

  std::optional<double> SampleDepth(const cv::Mat & depth, const std::pair<int, int> & centroid) {
    const int window = get_parameter("depth_window").as_int();
    const int half = std::max(1, window / 2);
    const int u = centroid.first;
    const int v = centroid.second;
    const int h = depth.rows;
    const int w = depth.cols;
    const int u_min = std::max(0, u - half);
    const int u_max = std::min(w, u + half + 1);
    const int v_min = std::max(0, v - half);
    const int v_max = std::min(h, v + half + 1);
    if (u_min >= u_max || v_min >= v_max) {
      return std::nullopt;
    }

    cv::Mat patch = depth(cv::Range(v_min, v_max), cv::Range(u_min, u_max));
    cv::Mat patch_float;
    if (depth.type() == CV_16U) {
      patch.convertTo(patch_float, CV_32F, 1.0 / 1000.0);
    } else {
      patch.convertTo(patch_float, CV_32F);
    }

    std::vector<float> values;
    if (patch_float.isContinuous()) {
      values.assign(
        reinterpret_cast<const float *>(patch_float.datastart),
        reinterpret_cast<const float *>(patch_float.dataend));
    } else {
      values.reserve(static_cast<size_t>(patch_float.total()));
      for (int row = 0; row < patch_float.rows; ++row) {
        const float * row_ptr = patch_float.ptr<float>(row);
        values.insert(values.end(), row_ptr, row_ptr + patch_float.cols);
      }
    }
    return MedianOf(values);
  }

  std::optional<std::array<double, 3>> ProjectToCamera(
    const std::pair<int, int> & centroid,
    double depth_m) const {
    if (!camera_info_) {
      return std::nullopt;
    }
    const int u = centroid.first;
    const int v = centroid.second;
    const auto & k = camera_info_->k;
    const double fx = k[0];
    const double fy = k[4];
    const double cx = k[2];
    const double cy = k[5];
    if (fx == 0.0 || fy == 0.0) {
      return std::nullopt;
    }
    const double x = (static_cast<double>(u) - cx) / fx * depth_m;
    const double y = (static_cast<double>(v) - cy) / fy * depth_m;
    const double z = depth_m;
    return std::array<double, 3>{x, y, z};
  }

  std::optional<std::array<double, 3>> CameraToWorld(
    const std::array<double, 3> & camera_point,
    const builtin_interfaces::msg::Time & stamp) {
    const auto body_frame = get_parameter("body_frame").as_string();
    const auto offset = get_parameter("camera_offset").as_double_array();
    const auto body_point = CameraToBody(camera_point);

    geometry_msgs::msg::PoseStamped pose;
    pose.header.frame_id = body_frame;
    pose.header.stamp = stamp;
    pose.pose.position.x = body_point[0] + offset.at(0);
    pose.pose.position.y = body_point[1] + offset.at(1);
    pose.pose.position.z = body_point[2] + offset.at(2);
    pose.pose.orientation.w = 1.0;

    const double timeout_s = get_parameter("tf_timeout_s").as_double();
    const bool use_latest = get_parameter("use_latest_tf_on_extrapolation").as_bool();
    auto world_pose = TransformPose(pose, timeout_s, use_latest);
    if (!world_pose) {
      return std::nullopt;
    }

    return std::array<double, 3>{
      world_pose->pose.position.x,
      world_pose->pose.position.y,
      world_pose->pose.position.z};
  }

  static std::array<double, 3> CameraToBody(const std::array<double, 3> & camera_point) {
    return {camera_point[2], -camera_point[0], -camera_point[1]};
  }

  std::optional<geometry_msgs::msg::PoseStamped> TransformPose(
    const geometry_msgs::msg::PoseStamped & pose,
    double timeout_s,
    bool use_latest) {
    const auto target_frame = get_parameter("world_frame").as_string();
    try {
      return tf_buffer_->transform(pose, target_frame, tf2::durationFromSec(timeout_s));
    } catch (const tf2::TransformException & exc) {
      const auto message = ToLower(exc.what());
      if (use_latest && message.find("extrapolation into the future") != std::string::npos) {
        auto latest_pose = pose;
        try {
          const auto latest_transform = tf_buffer_->lookupTransform(
            target_frame, pose.header.frame_id, tf2::TimePointZero);
          const rclcpp::Time latest_ros_time(latest_transform.header.stamp);
          const rclcpp::Time stamp_time(pose.header.stamp);
          if (latest_ros_time < stamp_time) {
            latest_pose.header.stamp = latest_transform.header.stamp;
          }
        } catch (const tf2::TransformException & lookup_exc) {
          RCLCPP_WARN(get_logger(), "TF latest transform lookup failed: %s", lookup_exc.what());
        }
        try {
          return tf_buffer_->transform(latest_pose, target_frame, tf2::durationFromSec(timeout_s));
        } catch (const tf2::TransformException & retry_exc) {
          RCLCPP_WARN(get_logger(), "TF transform failed: %s", retry_exc.what());
          return std::nullopt;
        }
      }
      RCLCPP_WARN(get_logger(), "TF transform failed: %s", exc.what());
      return std::nullopt;
    }
  }

  void UpdateTracks(const std::array<double, 3> & world_point) {
    const double gating_distance = get_parameter("gating_distance").as_double();
    Track * best_track = nullptr;
    double best_distance = 0.0;
    for (auto & track : tracks_) {
      const double dx = track.position[0] - world_point[0];
      const double dy = track.position[1] - world_point[1];
      const double dz = track.position[2] - world_point[2];
      const double distance = std::sqrt(dx * dx + dy * dy + dz * dz);
      if (distance <= gating_distance && (!best_track || distance < best_distance)) {
        best_track = &track;
        best_distance = distance;
      }
    }

    if (!best_track) {
      Track new_track;
      new_track.position = world_point;
      new_track.observations = 1;
      tracks_.push_back(new_track);
      return;
    }

    const int count = best_track->observations;
    best_track->position = {
      (best_track->position[0] * count + world_point[0]) / (count + 1),
      (best_track->position[1] * count + world_point[1]) / (count + 1),
      (best_track->position[2] * count + world_point[2]) / (count + 1)};
    best_track->observations += 1;
  }

  void PublishTracks(const builtin_interfaces::msg::Time & stamp) {
    const int min_observations = get_parameter("min_observations").as_int();
    const auto world_frame = get_parameter("world_frame").as_string();
    for (auto & track : tracks_) {
      if (track.observations < min_observations || track.published) {
        continue;
      }
      geometry_msgs::msg::PoseStamped pose;
      pose.header.stamp = stamp;
      pose.header.frame_id = world_frame;
      pose.pose.position.x = track.position[0];
      pose.pose.position.y = track.position[1];
      pose.pose.position.z = track.position[2];
      pose.pose.orientation.w = 1.0;
      detections_pub_->publish(pose);
      track.published = true;
    }
  }

  message_filters::Subscriber<sensor_msgs::msg::Image> semantic_sub_;
  message_filters::Subscriber<sensor_msgs::msg::Image> depth_sub_;
  message_filters::Subscriber<sensor_msgs::msg::CameraInfo> info_sub_;
  std::shared_ptr<Sync> sync_;

  sensor_msgs::msg::CameraInfo::ConstSharedPtr camera_info_;
  std::vector<Track> tracks_;

  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr detections_pub_;
};

int main(int argc, char ** argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<LanternDetectorNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
